File: src/lib/terminalSmart.ts
// ── Smart-mode Tauri wrappers ─────────────────────────────────────
// Typed `invoke()` wrappers for the smart-mode commands surfaced by
// `src-tauri/src/terminal_smart.rs`. Lives in its own file so M3..M6
// additions (completions, history queries, man-page summaries) don't
// pollute the main commands.ts.
//
// Per CLAUDE.md Rule 4, panels never call `invoke()` directly — they
// go through these helpers, and these helpers are the only place
// command names appear as raw strings.

import { invoke } from "@tauri-apps/api/core";

/** What `terminal_validate_command` returns. `kind` discriminates:
 *  - `"builtin"` — POSIX/bash/zsh shell builtin (cd, echo, …)
 *  - `"binary"` — found on `$PATH`; `path` is the absolute path
 *  - `"missing"` — not a builtin, not on `$PATH`; `path` is null */
export type CommandValidation = {
  kind: "builtin" | "binary" | "missing";
  path: string | null;
};

/** Resolve a command name against shell builtins + `$PATH`. Used by
 *  the smart-mode syntax overlay to red-line typos. The frontend
 *  caches results per-session in `useTerminalSmartStore`, so each
 *  unique name only crosses the IPC boundary once. */
export const terminalValidateCommand = (name: string) =>
  invoke<CommandValidation>("terminal_validate_command", { name });

/** Discriminator for `Completion.kind`. */
export type CompletionKind =
  | "builtin"
  | "binary"
  | "file"
  | "directory"
  | "subcommand"
  | "option"
  | "history";

/** One row in the completion popover. `value` is the full text the
 *  UI should produce when this row is selected; `display` is what
 *  to show in the row's main label; `hint` is the optional muted
 *  right-side annotation (resolved binary path, etc.); `description`
 *  is the localized side-panel text from the bundled command
 *  library (only set for `subcommand` / `option` rows). */
export type Completion = {
  kind: CompletionKind;
  value: string;
  display: string;
  hint?: string | null;
  description?: string | null;
};

/** Tab-completion candidates for the input line at `cursor`.
 *  Caller passes the shell's last-known `cwd` so the file branch
 *  resolves the right directory; pass `null` to fall back to the
 *  Pier-X process cwd inside pier-core. `locale` drives the
 *  language of subcommand / option descriptions returned for known
 *  commands (docker / git / kubectl / npm / ssh in the bundled
 *  pack); pass the active i18n locale, e.g. `"zh-CN"`. */
export const terminalCompletions = (
  line: string,
  cursor: number,
  cwd: string | null,
  locale: string | null = null,
) =>
  invoke<Completion[]>("terminal_completions", { line, cursor, cwd, locale });

/** A single option flag + its summary parsed from the man / --help
 *  output. Rendered as one row in the man-popover OPTIONS section. */
export type ManOption = {
  flag: string;
  summary: string;
};

/** Parsed man-page (or `--help` fallback) summary for one command. */
export type ManSynopsis = {
  synopsis: string;
  description: string;
  options: ManOption[];
  /** `"man"` when the data came from `man -P cat <cmd>`,
   *  `"help"` for the `<cmd> --help` fallback, `""` when neither
   *  was available. The popover shows this as a small muted hint
   *  so the user knows whether they're reading the canonical man
   *  page or a synthesised fallback. */
  source: string;
};

/** Look up the man-page summary for `command`. Resolves to `null`
 *  when neither `man` nor `--help` produced usable text — the
 *  popover renders an explicit "no documentation" message instead
 *  of treating that as an error. Genuine errors (invalid name, I/O
 *  failure) come back as a rejected promise. */
export const terminalManSynopsis = (command: string) =>
  invoke<ManSynopsis | null>("terminal_man_synopsis", { command });

/** Load `shell`'s persisted history file from disk. Resolves to
 *  `[]` for both "no file yet" and "no platform data dir"; the
 *  frontend ring then keeps the in-memory-only behaviour. */
export const terminalHistoryLoad = (shell: string) =>
  invoke<string[]>("terminal_history_load", { shell });

/** Append `command` to `shell`'s persisted history file. Backend
 *  silently drops lines that match the credential-keyword filter,
 *  so callers can fire-and-forget. */
export const terminalHistoryPush = (shell: string, command: string) =>
  invoke<void>("terminal_history_push", { shell, command });

/** Wipe `shell`'s persisted history file. Idempotent. */
export const terminalHistoryClear = (shell: string) =>
  invoke<void>("terminal_history_clear", { shell });

// ── Smart-mode command library ───────────────────────────────────
//
// Settings → Terminal → Command library. The library is a
// structured catalogue of commands with subcommand + option
// descriptions feeding the Tab completion popover. A small set
// ships bundled in the binary; users can install or update extra
// packs from disk (Phase D / E).

/** One row in the Settings library list. */
export type LibraryEntry = {
  command: string;
  toolVersion: string;
  /** `"bundled-seed"` (compiled-in), `"auto-imported"` (importer
   *  produced this from a CLI's `--help`/man/completion script),
   *  or `"user"` (hand-curated). */
  source: string;
  /** `"completion-zsh"` / `"man"` / `"help"` / `"hand-curated"`. */
  importMethod: string;
  /** `YYYY-MM-DD`. */
  importDate: string;
  subcommandCount: number;
  optionCount: number;
  /** Sorted list of locales present somewhere in the pack
   *  (e.g. `["en", "zh-CN"]`). */
  locales: string[];
};

/** Snapshot returned by the library Tauri commands. */
export type LibrarySnapshot = {
  entries: LibraryEntry[];
  /** Absolute path to the user pack directory; empty when the
   *  platform doesn't have an `app_data_dir`. */
  userDir: string;
};

/** List every loaded pack (bundled + user) for the Settings UI. */
export const completionLibraryList = () =>
  invoke<LibrarySnapshot>("completion_library_list");

/** Re-read user packs from disk and return the fresh snapshot. */
export const completionLibraryReload = () =>
  invoke<LibrarySnapshot>("completion_library_reload");

/** Install (or replace) a user pack. `body` is the raw JSON of a
 *  `CommandPack`; the backend validates schema + safe filename. */
export const completionLibraryInstallPack = (body: string) =>
  invoke<LibrarySnapshot>("completion_library_install_pack", { body });

/** Install a pack by reading the JSON file at the given absolute
 *  path. Returns the post-install snapshot. Used by the Settings
 *  "Import…" file picker so the frontend doesn't need to pull in a
 *  filesystem plugin just to forward the bytes. */
export const completionLibraryInstallPackFromPath = (path: string) =>
  invoke<LibrarySnapshot>("completion_library_install_pack_from_path", { path });

/** Remove a user pack by command name. Bundled packs are
 *  immutable; the UI hides the button for them. */
export const completionLibraryRemovePack = (command: string) =>
  invoke<LibrarySnapshot>("completion_library_remove_pack", { command });

File: src-tauri/src/terminal_smart.rs
//! Smart-mode Tauri commands.
//!
//! Thin wrappers around `pier_core::terminal::*` smart-mode pieces.
//! Kept in a sibling module rather than directly in `lib.rs` so the
//! M3..M6 surface (validation, completions, history, man-page
//! summaries) doesn't bloat the already-large root command file.
//!
//! Pure-IPC layer — every business-logic decision belongs in
//! `pier-core`. The shapes here just (de)serialise and forward.

use std::sync::OnceLock;

use pier_core::terminal::{
    complete_with_library, history_append, history_clear, history_load, man_synopsis,
    validate_command, CommandKind, Completion, Library, ManSynopsis,
};
use serde::Serialize;

/// Process-global command library. Populated lazily on first call
/// to [`completion_library`] from the bundled JSON packs. Phase C
/// will widen this into a `RwLock<Library>` so user packs and
/// online updates can replace it without restart; for now it's a
/// read-only cell — the library only contains compile-time data.
static COMPLETION_LIBRARY: OnceLock<Library> = OnceLock::new();

fn completion_library() -> &'static Library {
    COMPLETION_LIBRARY.get_or_init(Library::bundled)
}

/// Result of [`terminal_validate_command`].
///
/// `kind` is one of `"builtin"` / `"binary"` / `"missing"` so the
/// frontend can branch on a discriminator without rebuilding the
/// Rust enum on the TS side. `path` carries the absolute resolved
/// binary path when `kind == "binary"`, `null` otherwise.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommandValidation {
    pub kind: &'static str,
    pub path: Option<String>,
}

/// Resolve `name` against shell builtins + `$PATH`.
///
/// Called by the smart-mode syntax overlay each time it sees a new
/// command token in the user's currently-typed line. The frontend
/// caches results in a per-session LRU so a name only crosses the
/// IPC boundary once per session.
#[tauri::command]
pub fn terminal_validate_command(name: String) -> CommandValidation {
    match validate_command(&name) {
        CommandKind::Builtin => CommandValidation {
            kind: "builtin",
            path: None,
        },
        CommandKind::Binary(p) => CommandValidation {
            kind: "binary",
            path: Some(p.to_string_lossy().into_owned()),
        },
        CommandKind::Missing => CommandValidation {
            kind: "missing",
            path: None,
        },
    }
}

/// Tab-completion candidates for the input line at `cursor`.
///
/// Stateless — the caller passes the shell's last-known cwd (from
/// `terminal_current_cwd`) so this command doesn't need access to
/// `AppState`. Returning everything in one shot also keeps the IPC
/// path simple; the popover filters as the user types without
/// re-invoking until they hit Tab again.
///
/// `locale` selects the description language emitted by library-
/// driven rows (subcommands / option flags). Frontend passes the
/// active i18n locale (e.g. `"zh-CN"`); fallback chain inside the
/// library is `locale → language root → en → empty`.
#[tauri::command]
pub fn terminal_completions(
    line: String,
    cursor: usize,
    cwd: Option<String>,
    locale: Option<String>,
) -> Vec<Completion> {
    let cwd_path = cwd.as_deref().map(std::path::Path::new);
    let locale_str = locale.as_deref().unwrap_or("en");
    complete_with_library(&line, cursor, cwd_path, completion_library(), locale_str)
}

/// Look up the man-page summary (or `--help` fallback) for `cmd`.
///
/// Returns `Ok(None)` for the "no entry / no --help output" case so
/// the frontend can render an explicit "No documentation found"
/// message instead of treating it as a hard error. Genuine errors
/// (invalid name, I/O failure) come back as `Err(String)` and are
/// surfaced as toasts.
#[tauri::command]
pub fn terminal_man_synopsis(command: String) -> Result<Option<ManSynopsis>, String> {
    use pier_core::terminal::ManError;
    match man_synopsis(&command) {
        Ok(syn) => Ok(Some(syn)),
        Err(ManError::NotFound(_)) => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}

/// Load the persisted command-history ring for `shell` from disk.
/// Returns `Ok(vec![])` for either "no file yet" or "no usable
/// data dir on this platform" so the caller fails soft and falls
/// back to an in-memory-only history.
#[tauri::command]
pub fn terminal_history_load(shell: String) -> Result<Vec<String>, String> {
    use pier_core::terminal::HistoryError;
    match history_load(&shell) {
        Ok(v) => Ok(v),
        Err(HistoryError::NoDataDir) => Ok(Vec::new()),
        Err(e) => Err(e.to_string()),
    }
}

/// Append `command` to `shell`'s persisted history file. Drops the
/// line silently if it matches the credential-keyword filter (see
/// `pier_core::terminal::history::is_sensitive`); the in-memory
/// ring on the frontend still keeps it for the current session.
#[tauri::command]
pub fn terminal_history_push(shell: String, command: String) -> Result<(), String> {
    use pier_core::terminal::HistoryError;
    match history_append(&shell, &command) {
        Ok(()) => Ok(()),
        Err(HistoryError::NoDataDir) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}

/// Wipe the persisted history file for `shell`. Settings exposes
/// this through a "Clear history for this shell" button so the
/// user can purge a leaked entry without having to find the file
/// on disk.
#[tauri::command]
pub fn terminal_history_clear(shell: String) -> Result<(), String> {
    use pier_core::terminal::HistoryError;
    match history_clear(&shell) {
        Ok(()) => Ok(()),
        Err(HistoryError::NoDataDir) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}
